add license active ping

phritz opened the pull request.

phritz committed:
add license active ping

phritz commented:
Any objections to merging this?

aboodman commented on src/replicache.ts:
```diff
@@ -527,6 +583,16 @@ export class Replicache<MD extends MutatorDefs = {}> {
+   * for. We do not send pings when licensing is disabled (license key is
+   * undefined) or for the TEST_LICENSE_KEY.
+   */
+  get licenseActive(): Promise<boolean> {
```
What is the use case for this property?

phritz replied:
Currently this signal serves for testing: we can check assertions about active once this thing resolves. More generally I was just following the pattern established to signal that an asynchronous thing kicked off in the constructor has completed.

aboodman replied:
I'm worried about this becoming public API, which we would not want because we'd then be stuck supporting it even though it only exists for testing. Maybe there is some build mechanism I'm not aware of that prevents this and so my worry is invalid? Or maybe there's some way to mark the APi so it doesn't show up in docs as public but is still available on the class for testing?

Another way to control the asynchrony of the tests would be to use `sinon` to override the timers under test and force a number of ms to move forward and for timers to go off. Look for heartbeat.test.ts for example. I think this might be a win for complexity too as it would mean not having to wire these resolvers through the call stack and not having to track the extra state.

phritz replied:
ffwding timers does not solve the same problem: it can trigger timers but it doesn't tell us when an already-initiated call has completed. 

i understand the concern tho about putting it in the public api. i think the cleanest thing might be to make the accessor private and then in tests that need it add a mixin to replicache that promotes the call to public. is that a thing we do in javascript? if not then other options:
- an accessor for the license key could be a natural part of the api. if we had such an accessor we could gate it on the underlying license check and active ping calls, maybe including those bits along with the key in the return? 
- it could be hidden from generated jsdoc using `@ignore`
- make the accessor protected, inherit from repliache in the test that needs it, and promote the method to public for the test

aboodman commented on src/replicache.ts:
```diff
@@ -447,6 +466,42 @@ export class Replicache<MD extends MutatorDefs = {}> {
+      }
+    };
+    await activate();
+    resolveLicenseActive(true);
```
Whether the license is *active* is only a function of the license key which is known at construction time. Why use the promise/resolver mechanism to delay computing that? Seems more straightforward to compute it in the getter based on the license key.

phritz replied:
> Whether the license is active is only a function of the license key which is known at construction time.

There is a subtlety to it that means that no it is not. The definition of active given on the getter is _one that we **are currently sending** pings on_. The license isn't 'active' meaning activated until we actually start sending pings. Dunno if that is too subtle a distinction but that definition of 'active' is a more useful concept than 'we intend to send pings' because we can gate stuff that should only happen after license activation on it.

Maybe 'activated' would have been better?

aboodman commented on src/replicache.ts:
```diff
@@ -447,6 +466,42 @@ export class Replicache<MD extends MutatorDefs = {}> {
+      return noop;
+    }
+
+    const activate = async () => {
```
The use of the word "active" here seems to be overloading the definition of "active" as a license that requires pings. Why not call this function `pingLicense` or something?

phritz replied:
Two questions here I think.

1. "Why is this not called ping?" When first discussed, 'ping' was intended for something to do with keeping websocket connections alive, so I said I'd pick something else to avoid confusion. If 'ping' is no longer claimed, i would love to call this thing a ping.

2. "If called active, why overload that term?" The definition of 'active' I'm using means a license we are currently sending pings on, which I think is consistent with our use of "monthly active browsers" and similar. Is that overloading or diverging from some other concept of active?

aboodman replied:
Ah yes, ping is still used to eep the socket alive in Reflect.

OK I just had a hard time parsing this. The word "activate" typically means something you do once, not recurring. Maybe *markActive*?

aboodman replied:
I might even say `markLicenseActive`. I usually prefer your style of defining names up front to encapsulate concepts and decrease verbosity, but in this case I just can't think of a good one :).

phritz replied:
yeah neither could i. markactive seems workable. any chance you want to call the thingy that keeps sockets alive 'marco polo' or 'keepopen ack' or 'yo supdog' or something?

i will do markactive once i get an answer on whether i can use mixins

phritz reviewed (commented):
ok i tried my hand at a mixin. the tradeoff is it requires like 6 lines where we circumvent type safety. i'm sure arv or you could get it down further. in any case, better?

aboodman commented:
I think there is a typedoc annotation that makes public fields not
documented. We can consider such undocumented fields non-public api. Can
talk about this more next week.

On Fri, Mar 4, 2022 at 9:30 PM Phritz ***@***.***> wrote:

> ***@***.**** commented on this pull request.
>
> ok i tried my hand at a mixin. the tradeoff is it requires like 6 lines
> where we circumvent type safety. i'm sure arv or you could get it down
> further. in any case, better?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rocicorp/replicache/pull/851#pullrequestreview-900926229>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAATUBH64K2IZM4LU5YEQ2LU6METBANCNFSM5PWD5QTA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>
-- 
a (phone)

aboodman commented:
But I also sort of feel like there should be a way to test without exposing
this promise in the first place. Kind of interested to work through it if
you want to next week.

On Sat, Mar 5, 2022 at 10:10 AM Aaron Boodman ***@***.***>
wrote:

> I think there is a typedoc annotation that makes public fields not
> documented. We can consider such undocumented fields non-public api. Can
> talk about this more next week.
>
> On Fri, Mar 4, 2022 at 9:30 PM Phritz ***@***.***> wrote:
>
>> ***@***.**** commented on this pull request.
>>
>> ok i tried my hand at a mixin. the tradeoff is it requires like 6 lines
>> where we circumvent type safety. i'm sure arv or you could get it down
>> further. in any case, better?
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/rocicorp/replicache/pull/851#pullrequestreview-900926229>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAATUBH64K2IZM4LU5YEQ2LU6METBANCNFSM5PWD5QTA>
>> .
>> Triage notifications on the go with GitHub Mobile for iOS
>> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
>> or Android
>> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>>
>> You are receiving this because your review was requested.Message ID:
>> ***@***.***>
>>
> --
> a (phone)
>
-- 
a (phone)

phritz commented:
> I think there is a typedoc annotation that makes public fields not
documented.

Yeah, I offered @ignore as an alternative in https://github.com/rocicorp/replicache/pull/851#discussion_r819988437 

> But I also sort of feel like there should be a way to test without exposing this promise in the first place. 

Yes, I agree it should not be part of the public api, so per earlier comment I moved it to a private accessor that is "exposed" only in a test via a mixin. I don't see how we can await an async function without "exposiing" the underlying promise to test code in some fashion, so if the mixin isn't desirable I'll just move the promise to a free unexported function and test that. This is not a hard problem.

>  Kind of interested to work through it if you want to next week.

I be honest I'm not. I'm interested in getting something that works and doesn't mess up the api and moving on. The accessor is out of the api. If we don't like the mixin mechanism, I'll just move it to a free function. The only reason I added the promise via an accessor in the first place is because this is a well-established pattern in Replicache, there are three or four other promises exposed this way and it is obviously convenient for testing. They are proper parts of the public api tho, so they make more sense.

phritz commented:
Also the more I think about it the more it doesn't seem crazy to have a bit about the license in the public API. It's a first-class concept for us that there is a license also whether it is valid or not and also whether or not we are sending pings. But it's clearly not super useful so the default is not putting it in the public API I guess is a good one.

aboodman commented:
Hey, looking through this it looks like I missed a few statements/questions when I was quickly giving driveby feedback. Sorry about that. Circling back:

> Any objections to merging this?

Sorry I missed this question in the PR. I hope you were not waiting on my feedback, or interpreting my comments as objections.

As usual I'm fine landing stuff and reviewing async. There's almost nothing I think would be worth blocking landing on -- except things that break automated tests.

Even the API stuff, we do scan the commit messages for changes to the public API at release time, but I just realized we could have a more foolproof way to catch accidentally changes: diffing replicache.d.ts the auto-generated typescript typings.

So there's pretty much no reason to ever have a code review block landing. It's just too hard in a distributed setting.

> ffwding timers does not solve the same problem: it can trigger timers but it doesn't tell us when an already-initiated call has completed.

What I had in mind was using sinon's `useFakeTimers` and then rolling time forward to when the first license-related http request should happen, then using `fetchMock.flush()` to wait for them to complete. This is not as precise as using promises, but I think should work correctly and would reduce the plumbing of promises through a few functions and exposing state that feels internal out for the purposes of testing. It's not a big deal, and I'm not married to the idea.

> i think the cleanest thing might be to make the accessor private and then in tests that need it add a mixin to replicache that promotes the call to public. is that a thing we do in javascript?

No, mixins are not commonly used in JavaScript. Preference to avoid this (there's a very strong preference to avoid monkey-patching in modern JS).

However, your change made me realize that the simplest possible solution is just TypeScript's version of `reinterpret_cast`:

```ts
await (rep as unknown as {_fooPrivate: Promise<void>})._fooPromise;
```

> an accessor for the license key could be a natural part of the api. if we had such an accessor we could gate it on the underlying license check and active ping calls, maybe including those bits along with the key in the return?

I would not want to add any public API until a user asks. It's just so expensive to maintain and creates lots of extra points in the design space for future problems.

> it could be hidden from generated jsdoc using @ignore

I'm sorry I missed this when you mentioned it. I am fine with this, and it's a quick solution. However I think minor preference for something that doesn't add the public getter at all. I'm not sure where this preference is coming from. I have some spidey sense that `@ignore` might not be sufficient for keeping something out of our public API. I guess it just seems like a special case I'd prefer to not have to remember (we don't appear to use `@ignore` anywhere in Replicache yet).

> make the accessor protected, inherit from repliache in the test that needs it, and promote the method to public for the test

This seems like the best solution other than fake timers to me. You don't need to the accessor (getter) in Replicache at all -- you can remove the getter from `Replicache` completely and make the `_licenseActive` field `protected`. Then the subclass can access and expose as a public getter.

Summary -- I think the subclass is the best easy option.

arv commented:
`@ignore` is used when generating docs. It is not used by TypeScript.

In the past we have used `protected` and then expose a public version of the same in the [ReplicacheTest class](https://github.com/rocicorp/replicache/blob/main/src/test-util.ts#L28). Then we have [rules to not include private and protected](https://github.com/rocicorp/replicache/blob/main/doc/docusaurus.config.js#L21-L22) in the generated docs.

arv commented on src/replicache.test.ts:
```diff
@@ -2053,6 +2059,71 @@ test('licensing key is valid if check returns non-200', async () => {
+
+async function licenseActiveTest(tc: LicenseActiveTestCase) {
+  class MakeLicenseActivePublic {
+    public licenseActive(): Promise<boolean> {
```
public is default.

Maybe add an eslint rule to disallow this modifier?

arv commented on src/replicache.test.ts:
```diff
@@ -2053,6 +2059,71 @@ test('licensing key is valid if check returns non-200', async () => {
+  class MakeLicenseActivePublic {
+    public licenseActive(): Promise<boolean> {
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      return (this as any)._licenseActive();
```
Or use `// @ts-expect-error`

arv replied:
The benefit is that it would catch errors if the private method name changed.

arv commented on src/replicache.test.ts:
```diff
@@ -2053,6 +2059,71 @@ test('licensing key is valid if check returns non-200', async () => {
+  if (tc.expectFetchCalled) {
+    fetchMock.postOnce(activeUrlMatcher, tc.mockFetchParams);
+  }
+  mixin(Replicache, MakeLicenseActivePublic);
```
I see what you are doing here.... Just expose this on the ReplicacheTest class instead.

arv commented on src/replicache.ts:
```diff
@@ -527,6 +583,21 @@ export class Replicache<MD extends MutatorDefs = {}> {
+  // way for ts to know about it, hence:
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-expect-error
+  private _licenseActive(): Promise<boolean> {
```
Maybe remove the method and only have a method on the ReplicacheTest class.

arv reviewed (commented):
Lots of contention in a few lines.

Naming is hard...
Testing internal stuff is hard...

phritz commented:
Thanks. I went the subclass route via ReplicacheTest.
